File: src/lib/compact.ts
/**
 * Compact Context API client — DCP-inspired model-driven compaction.
 */

import { PHAOS_BASE } from './config';

export const compactApi = {
  getTokenCount: async (conversationId: string, modelName = 'default') => {
    const res = await fetch(`${PHAOS_BASE}/api/compact/conversation/tokens`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ conversation_id: conversationId, model_name: modelName }),
    });
    if (!res.ok) throw new Error(`Failed to get tokens: ${res.statusText}`);
    return res.json();
  },

  compactConversation: async (conversationId: string, start: number, end: number, summary: string, modelName = 'default') => {
    const res = await fetch(`${PHAOS_BASE}/api/compact/conversation/compact`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ conversation_id: conversationId, start, end, summary, model_name: modelName }),
    });
    if (!res.ok) throw new Error(`Failed to compact: ${res.statusText}`);
    return res.json();
  },

  forceCompact: async (conversationId: string, modelName = 'default') => {
    const res = await fetch(`${PHAOS_BASE}/api/compact/conversation/tokens`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ conversation_id: conversationId, model_name: modelName }),
    });
    if (!res.ok) throw new Error(`Failed to get tokens: ${res.statusText}`);
    const usage = await res.json();
    if (usage.total > 0) {
      const compactRes = await fetch(`${PHAOS_BASE}/api/compact/conversation/compact`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          conversation_id: conversationId,
          start: 0,
          end: Math.floor(usage.total / (usage.max * 0.01)),
          summary: 'Context compacted by user request.',
          model_name: modelName,
        }),
      });
      if (!compactRes.ok) throw new Error(`Failed to compact: ${compactRes.statusText}`);
      return compactRes.json();
    }
    return { success: true, original_count: 0, compacted_count: 0 };
  },

  discardMessages: async (conversationId: string, messageIds: number[], modelName = 'default') => {
    const res = await fetch(`${PHAOS_BASE}/api/compact/conversation/discard`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ conversation_id: conversationId, message_ids: messageIds, model_name: modelName }),
    });
    if (!res.ok) throw new Error(`Failed to discard: ${res.statusText}`);
    return res.json();
  },

  protectMessages: async (conversationId: string, messageIds: number[], modelName = 'default') => {
    const res = await fetch(`${PHAOS_BASE}/api/compact/conversation/protect`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ conversation_id: conversationId, message_ids: messageIds, model_name: modelName }),
    });
    if (!res.ok) throw new Error(`Failed to protect: ${res.statusText}`);
    return res.json();
  },

  pruneToolCalls: async (conversationId: string, toolCallIds: string[], modelName = 'default') => {
    const res = await fetch(`${PHAOS_BASE}/api/compact/conversation/prune`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ conversation_id: conversationId, tool_call_ids: toolCallIds, model_name: modelName }),
    });
    if (!res.ok) throw new Error(`Failed to prune: ${res.statusText}`);
    return res.json();
  },

  getTools: async () => {
    const res = await fetch(`${PHAOS_BASE}/api/compact/tools`);
    if (!res.ok) throw new Error(`Failed to get tools: ${res.statusText}`);
    return res.json();
  },
};

File: phaos/routes/compact.py
"""Compact Context API routes — DCP-inspired model-driven compaction."""

from __future__ import annotations

from fastapi import APIRouter
from pydantic import BaseModel
from typing import List, Optional

from ..core.compact_context import get_compact_context, COMPACT_TOOL_DEFINITIONS
from ..db.conversation_store import get_messages, delete_message, create_message

router = APIRouter(tags=["compact"])


def _persist_messages(conversation_id: str, new_messages: list[dict]):
    """Replace all messages for a conversation with new_messages."""
    old_messages = get_messages(conversation_id)
    for msg in old_messages:
        delete_message(msg["id"])
    for msg in new_messages:
        create_message(conversation_id, msg.get("role", "user"), msg.get("content", ""))


class TokenCountRequest(BaseModel):
    conversation_id: str
    model_name: str = "default"


class CompactRequest(BaseModel):
    conversation_id: str
    start: int
    end: int
    summary: str
    model_name: str = "default"


class DiscardRequest(BaseModel):
    conversation_id: str
    message_ids: List[int]
    model_name: str = "default"


class ProtectRequest(BaseModel):
    conversation_id: str
    message_ids: List[int]
    model_name: str = "default"


class PruneRequest(BaseModel):
    conversation_id: str
    tool_call_ids: List[str]
    model_name: str = "default"


@router.post("/conversation/tokens")
async def get_token_count(req: TokenCountRequest):
    messages = get_messages(req.conversation_id)
    ctx = get_compact_context(req.model_name)
    return ctx.get_usage(messages)


@router.post("/conversation/compact")
async def compact_conversation(req: CompactRequest):
    messages = get_messages(req.conversation_id)
    ctx = get_compact_context(req.model_name)
    result = ctx.compact(messages, start=req.start, end=req.end, summary=req.summary)
    _persist_messages(req.conversation_id, result)
    return {"success": True, "original_count": len(messages), "compacted_count": len(result)}


@router.post("/conversation/discard")
async def discard_messages(req: DiscardRequest):
    messages = get_messages(req.conversation_id)
    ctx = get_compact_context(req.model_name)
    result = ctx.discard(messages, message_ids=req.message_ids)
    _persist_messages(req.conversation_id, result)
    return {"success": True, "original_count": len(messages), "new_count": len(result)}


@router.post("/conversation/protect")
async def protect_messages(req: ProtectRequest):
    messages = get_messages(req.conversation_id)
    ctx = get_compact_context(req.model_name)
    ctx.protect(messages, message_ids=req.message_ids)
    return {"success": True, "protected": req.message_ids}


@router.post("/conversation/prune")
async def prune_messages(req: PruneRequest):
    messages = get_messages(req.conversation_id)
    ctx = get_compact_context(req.model_name)
    result = ctx.prune(messages, tool_call_ids=req.tool_call_ids)
    _persist_messages(req.conversation_id, result)
    return {"success": True, "original_count": len(messages), "new_count": len(result)}


@router.get("/tools")
async def get_compact_tools():
    return COMPACT_TOOL_DEFINITIONS
